src/geos.c

```diff
@@ -892,67 +894,205 @@ GEOSGeometry* PyGEOSForce3D(GEOSContextHandle_t ctx, GEOSGeometry* geom, double
+        return PGERR_GEOS_EXCEPTION;
+      }
+    }
+  }
+  return PGERR_SUCCESS;
+}
+
+char fill_coord_seq_ignore_nan(GEOSContextHandle_t ctx, GEOSCoordSequence* coord_seq,
+                               const double* buf, unsigned int dims, npy_intp cs1,
+                               npy_intp cs2, unsigned int first_i, unsigned int last_i) {
+  unsigned int i, j, current = 0;
+  char *cp1, *cp2;
+  char this_coord_is_finite;
+  double coord;
+
+  cp1 = (char*)buf + cs1 * first_i;
+  for (i = first_i; i <= last_i; i++, cp1 += cs1) {
+    cp2 = cp1;
+    this_coord_is_finite = 1;
+    for (j = 0; j < dims; j++, cp2 += cs2) {
+      coord = *(double*)cp2;
+      if (!npy_isfinite(coord)) {
+        this_coord_is_finite = 0;
+        break;
+      }
+      if (!GEOSCoordSeq_setOrdinate_r(ctx, coord_seq, current, j, coord)) {
+        return PGERR_GEOS_EXCEPTION;
+      }
+    }
+    if (this_coord_is_finite) {
+      current++;
+    }
+  }
+  return PGERR_SUCCESS;
+}
+
 /* Create a GEOSCoordSequence from an array
  *
  * Note: this function assumes that the dimension of the buffer is already
  * checked before calling this function, so the buffer and the dims argument
  * is only 2D or 3D.
+ *
+ * handle_nans: 0 means 'allow', 1 means 'ignore', 2 means 'raise'
+ *
+ * Returns an error state (PGERR_SUCCESS / PGERR_GEOS_EXCEPTION / PGERR_NAN_COORD).
  */
-GEOSCoordSequence* coordseq_from_buffer(GEOSContextHandle_t ctx, const double* buf,
-                                        unsigned int size, unsigned int dims, char ring_closure,
-                                        npy_intp cs1, npy_intp cs2) {
-  GEOSCoordSequence* coord_seq;
+int coordseq_from_buffer(GEOSContextHandle_t ctx, const double* buf, unsigned int size,
+                         unsigned int dims, char is_ring, int handle_nans, npy_intp cs1,
+                         npy_intp cs2, GEOSCoordSequence** coord_seq) {
   char *cp1, *cp2;
-  unsigned int i, j;
-  double first_coord;
+  unsigned int i, j, current, first_i, last_i, actual_size;
+  double coord;
+  char errstate;
+  char ring_closure = 0;
+
+  switch (handle_nans) {
+    case PYGEOS_HANDLE_NANS_ALLOW:
+      actual_size = size;
+      first_i = 0;
+      last_i = size - 1;
+      break;
+    case PYGEOS_HANDLE_NANS_IGNORE:
+      actual_size = count_finite(buf, size, dims, cs1, cs2, &first_i, &last_i);
+      break;
+    case PYGEOS_HANDLE_NANS_RAISE:
+      actual_size = count_finite(buf, size, dims, cs1, cs2, &first_i, &last_i);
+      if (actual_size != size) {
+        return PGERR_NAN_COORD;
+      }
+      break;
+    default:
+      return PGERR_NAN_COORD;
+  }
 
-#if GEOS_SINCE_3_10_0
+  if (actual_size == 0) {
+    *coord_seq = GEOSCoordSeq_create_r(ctx, 0, dims);
+    return (*coord_seq != NULL) ? PGERR_SUCCESS : PGERR_GEOS_EXCEPTION; 
+  }
+
+  /* Rings automatically get an extra (closing) coordinate if they have
+     only 3 or if the first and last are not equal. */
```

Comment:
As you can see from my earlier comment, I was initially surprised by this "always close if they have only 3". But this is the current behaviour of course and can be left for a separate issue. But we _could_ also only do that if the first/last are not equal?

Reply:
See comment above.